File: src/com/chbachman/cminus/representation/Type.java
package com.chbachman.cminus.representation;

import com.chbachman.cminus.CMinusParser;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by Chandler on 4/12/17.
 */
public class Type implements Typed {
    private static Map<String, Type> types = new HashMap<>();

    public static void init() {
        for (Native n : Native.values()) {
            types.put(n.type.type, n.type);
        }
    }

    public final String type;
    public final String cType;

    private Type(String type) {
        this(type, type);
    }

    private Type(String type, String cType) {
        this.type = type;
        this.cType = cType;
    }

    public String code() {
        return cType;
    }

    @Override
    public Type type() {
        return this;
    }

    public static Type from(CMinusParser.TypeContext context) {
        return getType(context.ID().getText());
    }

    public static Type from(Struct struct) {
        return getType(struct.name, "struct " + struct.name);
    }

    private static Type getType(String name) {
        return getType(name, name);
    }

    private static Type getType(String name, String cType) {
        if (types.containsKey(name)) {
            return types.get(name);
        } else {
            Type t = new Type(name, cType);
            types.put(name, t);
            return t;
        }
    }

    public enum Native {

        STRING("String", "char *"),
        INT("Int", "int"),
        VOID("Void", "void"),
        BOOL("Bool", "int");

        public Type type;

        Native(String type) {
            this(type, type);
        }

        Native(String type, String cType) {
            this.type = new Type(type, cType);
        }

        public static Native get(Type type) {
            for (Native n : Native.values()) {
                if (n.type == type) {
                    return n;
                }
            }

            return VOID;
        }

    }
}

File: src/com/chbachman/cminus/representation/Type.kt
package com.chbachman.cminus.representation

import com.chbachman.cminus.CMinusParser

import java.util.HashMap

class Type private constructor(val type: String, val cType: String = type) : Typed {

    fun code(): String {
        return cType
    }

    override fun type(): Type {
        return this
    }

    enum class Native constructor(type: String, cType: String = type) {

        STRING("String", "char *"),
        INT("Int", "int"),
        VOID("Void", "void"),
        BOOL("Bool", "int");

        var type = Type(type, cType)

        companion object {
            operator fun get(type: Type): Native {
                return Native.values().find { it.type === type } ?: VOID
            }
        }
    }

    companion object {
        private val types = HashMap<String, Type>()

        fun init() {
            for (n in Native.values()) {
                types.put(n.type.type, n.type)
            }
        }

        fun from(context: CMinusParser.TypeContext): Type {
            return getType(context.ID().text)
        }

        fun from(struct: Struct): Type {
            return getType(struct.name, "struct " + struct.name)
        }

        private fun getType(name: String, cType: String = name): Type {
            if (types.containsKey(name)) {
                return types[name]!!
            } else {
                val t = Type(name, cType)
                types.put(name, t)
                return t
            }
        }
    }
}
